### PR TITLE
Always report query result to query-frontend, even when cancelled

### PR DESCRIPTION
**Pull request by charleskorn**
#### What this PR does

This PR fixes an issue where queriers can fail to send query results to query-frontends if the query is cancelled for some reason. This manifests as query-frontends timing out waiting for query results.

This can happen in two scenarios:

* the query-scheduler dies or the query-frontend cancels the query, causing the gRPC stream between the querier and the query-scheduler to be closed and triggering cancellation of the query
* there's a race between when the querier cancels the contexts for processors for schedulers that are shutting down [here](https://github.com/grafana/mimir/blob/14e4b3eea2c67a44da1d8aa90c53d12299006398/pkg/querier/worker/processor_manager.go#L87-L90) and when the querier cancels the execution context created in `newExecutionContext` [here](https://github.com/grafana/mimir/blob/3a401cacc822fa25f542d45cb431248e05502610/pkg/querier/worker/util.go#L51) - it's possible that this happens:
  - `querierLoop` receives a query
  - `newExecutionContext` observes that `workerCtx` is cancelled and `inflightQuery` is `false` and cancels the execution context
  - `querierLoop` then sets `inflightQuery` to `true` and begins executing the query, which quickly fails because the context has been cancelled

https://github.com/grafana/mimir/pull/6706 fixes the second issue, but until that issue is fixed, this PR will at least mitigate the impact of that issue: the querier will report the query as cancelled and the query-frontend can retry the query, rather than the query-frontend simply timing out.

#### Which issue(s) this PR fixes or relates to

(none)

#### Checklist

- [x] Tests updated.
- [n/a] Documentation added.
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [n/a] [`about-versioning.md`](/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

File: pkg/querier/worker/scheduler_processor_test.go [hunk 1]
```diff
@@ -97,10 +100,63 @@ func TestSchedulerProcessor_processQueriesOnSingleStream(t *testing.T) {
...
+		workerCtx, workerCancel := context.WithCancel(context.Background())
+
+		// processQueriesOnSingleStream() blocks and retries until its context is cancelled, so run it in the background.
+		go func() {
+			sp.processQueriesOnSingleStream(workerCtx, nil, "127.0.0.1")
+		}()
+
+		require.Eventually(t, func() bool {
+			return frontend.queryResultCalls.Load() == 1
+		}, time.Second, 10*time.Millisecond, "expected frontend to be informed of query result exactly once")
```
Comment (jhesketh) [inline pkg/querier/worker/scheduler_processor_test.go:153]: (nit) Will this actually check if it is exactly once? It's possible that tick happens inbetween a result being returned where the conditional function returns true already.

Perhaps another assert after workCancel below to ensure it was only once?

**Review by jhesketh — COMMENTED** [review]
Do we also test that the query-frontend retries a query after receiving an canceled response back from a querier?

File: pkg/querier/worker/scheduler_processor_test.go (hunk 1 above)
Comment (charleskorn) [reply pkg/querier/worker/scheduler_processor_test.go:153]: The test sets up the fake scheduler to only ever return one query, so only one query result should ever be returned to the frontend. (Without the fix in this PR, the frontend received no results at all, so observing one result being returned indicates the issue has been fixed.)

If the querier was calling the frontend multiple times for some reason, then I'd expect this test to flake occasionally.

**Comment by charleskorn** [comment]
>Do we also test that the query-frontend retries a query after receiving an canceled response back from a querier?

Cancelled requests are reported back to the query-frontend as HTTP 500 errors, and the query-frontend [retries](https://github.com/grafana/mimir/blob/6eb5f34818b32c91d113101e7a6cb16d4842f3b0/pkg/frontend/querymiddleware/retry.go#L86) those. (I've confirmed this happens by checking the behaviour of one of Grafana Labs' internal Mimir clusters.)

If the query-frontend itself has cancelled the request, then it [won't retry](https://github.com/grafana/mimir/blob/6eb5f34818b32c91d113101e7a6cb16d4842f3b0/pkg/frontend/querymiddleware/retry.go#L81).

**Review by pracucci — APPROVED** [review]
I think the change is not wrong per se. I'm approving because I think it's a step forward, but I have some comments.

> This can happen in two scenarios:
> the query-scheduler dies

If the query-scheduler dies and the gRPC context is canceled, then the context used to run the query gets cancelled [here](https://github.com/grafana/mimir/blob/a5cc45a8596cdfed66fb88b3ef2727202edb6b78/pkg/querier/worker/scheduler_processor.go#L138-L140). However, it was just the query-scheduler to be terminated, so why cancelling the query execution? Couldn't the query execution complete and report the result to query-frontend anyway, instead of reporting the context cancellation?

I guess the hard problem here is to distinguish between a context canceled because query-scheduler died or because query was canceled by query-frontend. If my understanding is correct, then the current logic looks a good-enough trade-off.

> This can happen in two scenarios:
> the query-frontend cancels the query

If it's the query-frontend to cancel the query, what's the purpose of reporting back the "context canceled"? Can you confirm doing it for this case doesn't fix anything, it's just that we're going to do it even in this case, given we can't distinguish between a context canceled because of query-scheduler or query canceled?

**Review by dimitarvdimitrov — COMMENTED** [review]
I don't understand why the querier would think that the query was cancelled and the query-frontend would still wait for the response. The query-scheduler would cancel the stream if the query-frontend requested that.

Doesn't this indicate some bad context handling in the chain?

**Review by dimitarvdimitrov — APPROVED** [review]
I agree with Marco that this covers the query-scheduler ungracefully terminating and it's a step forward

**Comment by charleskorn** [comment]
> If the query-scheduler dies and the gRPC context is canceled, then the context used to run the query gets cancelled [here](https://github.com/grafana/mimir/blob/a5cc45a8596cdfed66fb88b3ef2727202edb6b78/pkg/querier/worker/scheduler_processor.go#L138-L140). However, it was just the query-scheduler to be terminated, so why cancelling the query execution? Couldn't the query execution complete and report the result to query-frontend anyway, instead of reporting the context cancellation?
> 
> I guess the hard problem here is to distinguish between a context canceled because query-scheduler died or because query was canceled by query-frontend. If my understanding is correct, then the current logic looks a good-enough trade-off.

Yep, I agree - it'd be great if we could distinguish between the two cases, but I think that will take more work, and I wanted to get a simple solution together to mitigate the issues we've been seeing and protect against other cases we might not have thought of. 

We might be able to return a different error from schedulers [here](https://github.com/grafana/mimir/blob/e238310c6063b642e4d4007fbba01f6ef1ff532d/pkg/scheduler/scheduler.go#L481) so that queriers can distinguish between the two cases. 

> 
> > This can happen in two scenarios:
> > the query-frontend cancels the query
> 
> If it's the query-frontend to cancel the query, what's the purpose of reporting back the "context canceled"? Can you confirm doing it for this case doesn't fix anything, it's just that we're going to do it even in this case, given we can't distinguish between a context canceled because of query-scheduler or query canceled?

Yep, exactly - this is just to keep things simple and make sure we don't miss reporting back to query-frontends in other cases.

**Pull request merged**